File: python/src/fintech_volume_imbalance_bars/tape.py
"""Trade-tape loader for Volume-Imbalance Bars.

Imbalance bars need each execution's **price** (to assign a tick sign) and its
**volume** (to weight that sign), so they are built from a trade tape. Yahoo
Finance does not expose tick data — only pre-aggregated bars, where the per-trade
signs are already lost — so there is no live Yahoo source for this algorithm.
"""

from __future__ import annotations

import csv
from pathlib import Path
from typing import Any

__all__ = ["load_trades"]

_TRADE_FIELDS = ("tradeId", "timestamp", "session", "symbol", "currency")


def load_trades(csv_path: str | Path) -> list[dict[str, Any]]:
    """Read a trade tape from CSV into trade dicts for :func:`construct_bars`.

    Expected columns: ``tradeId,timestamp,session,symbol,price,volume,currency``.
    ``price`` / ``volume`` become floats.
    """
    with open(csv_path, newline="", encoding="utf-8") as handle:
        rows = list(csv.DictReader(handle))
    trades: list[dict[str, Any]] = []
    for row in rows:
        trade: dict[str, Any] = {field: row[field] for field in _TRADE_FIELDS if field in row}
        trade["price"] = float(row["price"])
        trade["volume"] = float(row["volume"])
        trades.append(trade)
    return trades

File: typescript/src/tape.ts
/**
 * Trade-tape loader for Volume-Imbalance Bars.
 *
 * Imbalance bars need each execution's price (to assign a tick sign) and its
 * volume (to weight that sign), so they are built from a trade tape. Yahoo
 * Finance does not expose tick data — only pre-aggregated bars, where the
 * per-trade signs are already lost.
 */

import { readFileSync } from "node:fs";

import type { Trade } from "./bars.ts";

const TRADE_FIELDS = ["tradeId", "timestamp", "session", "symbol", "currency"] as const;

/** Read a trade tape from a CSV into trades for `constructBars`. */
export function loadTrades(csvPath: string): Trade[] {
  const text = readFileSync(csvPath, "utf8").trim();
  const lines = text.split(/\r?\n/);
  if (lines.length < 2) return [];
  const header = lines[0].split(",");
  return lines.slice(1).map((line) => {
    const cells = line.split(",");
    const row: Record<string, string> = {};
    header.forEach((name, i) => {
      row[name] = cells[i];
    });
    const trade: Record<string, unknown> = {};
    for (const field of TRADE_FIELDS) if (field in row) trade[field] = row[field];
    trade.price = Number(row.price);
    trade.volume = Number(row.volume);
    return trade as unknown as Trade;
  });
}
